Multiprocessing

Psykopear opened the pull request:
This PR removes the `multiprocessing`  and `dill` dependencies.

We used `multiprocessing` to spawn a cluster on the local machine, but its use was causing some problems, and we weren't using most of the library's functionalities.

With this PR instead of a `spawn_cluster` function we have a module that acts as a script to handle the processes in `bytewax.run`.

See `python -m bytewax.run -h` for all the possible options.

This allows us to remove the need for the whole dataflow to be pickable, and while we still need the ability to pickle python objects that are saved into recovery state, this should reduce the scope, so we decided to remove the `dill` dependency too, in favor of python's `pickle` module.

`run_main` and `cluster_main` can still be called where the dataflow is defined, but crash when the file is imported through `bytewax.run`, since the script will call either `run_main` or `cluster_main` depending on the options.

The hard work is done, so changing the user facing api shouldn't be too hard, and I'm more than open to suggestions and changes there.

davidselassie commented on pysrc/bytewax/execution.py:
```diff
@@ -3,16 +3,7 @@
-from bytewax.recovery import RecoveryConfig
-
-from .bytewax import cluster_main, run_main  # noqa: F401
+from .bytewax import cluster_main, run_main, spawn_cluster  # noqa: F401
```
Also I wouldn't mind moving all this directly into `bytewax.run` since it's all used from there. And is shorter to type. I don't think we need a separate mod.

Psykopear replied:
`cluster_main` is now in `bytewax.run`, while `run_main` is in `bytewax.testing` to make it clear it should be used in production.

davidselassie commented on pytests/test_execution.py:
```diff
@@ -46,73 +42,52 @@ def boom(item):
-)
-def test_output_parts_hashed_to_workers(mp_ctx):
-    pass
+def test_cluster_can_be_ctrl_c():
```
I think it makes sense to test `spawn_cluster` here directly. And then have some separate tests for the CLI wrapper function that's a little more light weight.

Psykopear replied:
There is a hidden detail that breaks this: since we need to respawn a new process with the same parameters multiple times, `spawn_cluster` under the hood calls the process with the same command line arguments. During a pytest run this means that spawn_cluster calls pytest again rather than "python -m bytewax.run", so if we want to test the script we need to test it "from outside". `spawn_cluster` should not be used directly outside of `bytewax.run`.

Another approach would be to force running "python -m bytewax.run" and manually rebuild the rest of the args  instead of reusing `sys.args` but this means we make some assumptions which might not always be true, like "python" being the correct executable. I'll look at this after the rest of the fixes

davidselassie reviewed (commented):
Thanks for working on this. I like the trick of importing a file as a dynamic module in each worker, and I like the `get_flow` builder interface. I know this is still a draft, but I wanted to get some feedback out there on the CLI parts and I'm sure you'll keep refining things. I think my biggest feedback is it feels a little confusing that sometimes you want to put `run_main` in the dataflow Python file itself, and sometimes you don't / it is ignored, but it's not uniform when. See my giant comment below about that.

Psykopear commented on pyproject.toml:
```diff
@@ -19,7 +19,7 @@ classifiers = [
     "Topic :: System :: Networking",
     "Topic :: System :: Distributed Computing",
 ]
-dependencies = ["multiprocess>=0.70", "dill>=0.3.5"]
```
Uh, right that was the whole point of removing multiprocessing. After I fix the rest of the issues I'll work on this

Psykopear replied:
fixed

Psykopear commented on pysrc/bytewax/inputs.py:
```diff
@@ -222,3 +222,17 @@ def build(self, worker_index, worker_count) -> StatelessSource:
         ...
+
+
+class EmptySource(StatelessSource):
```
Ook, it did feel like I was misusing this, I'm going to delete this and use the PartitionedOutput on the examples.

Psykopear commented on examples/errors/working.py:
```diff
@@ -25,13 +28,13 @@ def close(self):
 class NumberInput(DynamicInput):
```
Yes this "working" example here is just a placeholder that uses everything that can be broken in the errors examples, I'd leave it here as a reference and replace it in other error examples where I don't actually fail in the input

Psykopear commented on examples/errors/map_error.py:
```diff
@@ -4,10 +4,32 @@
+        pass
+
+
+class NumberInput(DynamicInput):
```
:+1:

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,100 @@
+    recovery.add_argument(
+        "-r", "--recovery-engine", type=str, choices=["kafka", "sqlite"]
+    )
+    kafka_config = parser.add_argument_group("Kafka recovery config")
```
Removing the option would make the cli nicer, but depending on how we go forward we might still need a way of specifying this here. See the answer to the big comment

Psykopear replied:
Removed in the meantime

Psykopear commented on src/execution/mod.rs:
```diff
@@ -687,6 +690,11 @@ pub(crate) fn run_main(
     epoch_interval: Option<EpochInterval>,
     recovery_config: Option<Py<RecoveryConfig>>,
 ) -> PyResult<()> {
+    if is_in_bytewax_run(py)? {
```
Yes, this was an idea I had yesterday to keep both the old execution methods and the new one, but I don't like it either. I opened this draft PR to get feedback on this, so thanks.

I did this to make the changes a little bit less breaking, but if we are ok on completely redefining the execution model things might get easier.

> Our core dataflow definition is always defined by a "dataflow builder" in a standalone Python file, like your existing get_flow function idea.

Ok, I like this since it avoids the double meaning of the "dataflow-name" argument, which is not good from an api point of view. So we require users to define a getter function.

> Our lowest level entry points like run_main and cluster_main don't know anything about modules, they use fully instantiated Dataflow objects and are maybe considered "private"? They are not meant to be called by users of the library. If people want to do freaky things with Jupyter notebooks, fine, but I think from here on out I'm going to consider that using "private interfaces".

So, I approached this trying to make it easier for existing users to upgrade, but it ended up confusing.
I did it this way because executing any of the running functions during the import leads to a non obvious behaviour: if you try to run with `bytewax.run` and recovery and multiple processes, you see the dataflow running, but it's not running with the config you want. But I agree that just ignoring the code is possibly even more confusing.
Rethinking this, we could move `run_main` (maybe renamed to `run_dataflow`) to `bytewax.testing`   so that it's clear it shouldn't be used in production, and keep the trick to detect if the script was ran with `bytewax.run` and, instead of ignoring the code, just crash the process asking to remove the call?

`cluster_main` still has its reasons for existing: if you want to run processes on different machines both the old and the new `spawn_cluster` wouldn't allow this. And if we remove the "kafka" recovery option from the script, this is the only entrypoint that would allow using it.

I think we either expand the `run` script to allow everything possible with `cluster_main`, and then remove `cluster_main` from the public api completely, or we specify that the `bytewax.run` script is only good for single machine dataflows, but we keep `cluster_main` public and documented for advanced cases.

Currently I'm leaning towards having less entrypoints to avoid confusion, so just `bytewax.testing.run_dataflow` and `bytewax.run`, but it means we have to add some more config options to the script.

> spawn_cluster sits on top of this and does instantiate dataflows from modules and uses the get_flow function and does the process coordination. We can use this in unit tests sometimes (writing those specific dataflows-under-test in a separate test file to be imported), but it isn't really public. (We can also still use run_main and cluster_main in tests too.)

Ok, this means we restrict multiprocess dataflow tests to those specific tests, but I don't think that's bad.

> On to of all this is the run module and python -m bytewax.run. This is the only "user facing" entry point for running dataflows. It looks at the command line flags and decides whether to use run_main, spawn_cluster, or cluster_main internally based on command line flags or env vars.

Which is more or less what we have here. With the latest commits from yesterday I changed things, and now `spawn_cluster` is only using `cluster_main`, but it previously used `run_main` by default and only went for `cluster_main` if specific parameters were passed. I changed it to make the trick of ignoring the code work in some special cases, but if we go for the crashing approach the problem should not be there.
 
> You've thought about this more than me. Perhaps there's some gotchas surrounding doing dynamic imports that makes this difficult.

There are, but the things you suggested mostly help with those, so it should be ok.

I'm going to work on the suggestions and update the PR so we can have a second round of review. I know we want this for the next release, but I don't want to rush this.

Psykopear commented:
Ok pushed some more commits.

- Removed `KafkaRecoveryConfig` option from the cli, so I left `cluster_main` publicly available since it's still the only way to use kafka recovery and to run processes on different machines.
- `run_main` is now in `bytewax.testing`.
- If `run_main` or `cluster_main` are called while importing the dataflow in `bytewax.run`/`spawn_cluster`, the process crashes telling the user to remove the line.
- `--dataflow-name` is now `--dataflow-builder`, and only accepts a function that returns a Dataflow object.
- fixed all the wrong usage of DynamicInput in the errors examples and wikistream.

Psykopear commented on src/errors.rs:
```diff
@@ -39,7 +42,26 @@ pub(crate) trait PythonException<T> {
+                //
+                // This means that our message will be quoted if we reraise
+                // it as it is. So in this case we raise a RuntimeError instead.
+                if err
+                    .get_type(py)
+                    .is(pyo3::types::PyType::new::<pyo3::exceptions::PyKeyError>(py))
+                {
+                    PyRuntimeError::new_err(build_message(py, caller, &err, msg))
```
This is another place where I do something "surprising", and I'm not sure if we want this.
Python special cases KeyError to quote the error message, so our nicely formatted error becomes a string with `\n` ecc.
To avoid this, I change the error to a PyRuntimeError, which "fixes" the problem with the error representation, but makes it impossible to catch any KeyError on the Python side. Which in general should be ok because if the dataflow run fails, we don't really want users to catch the exception and try again in the same process, but it still is a surpriing behaviour.

Psykopear commented:
Latest commit is a refactoring of the execution model. I ended up touching the recovery module too, which is a bit out of scope for this PR, but this way we follow the same builder pattern we use in other modules, and it makes the execution part clearer. If we don't like this, or want it in a separate PR, I can revert the commit.

Psykopear commented on examples/wikistream.py:
```diff
@@ -26,25 +33,12 @@ def close(self):
 
 class WikiStreamInput(DynamicInput):
```
fixed

davidselassie commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,89 @@
+    if sqlite_directory:
+        recovery_config = SqliteRecoveryConfig(sqlite_directory or "./")
+    kwargs["recovery_config"] = recovery_config
+    spawn_cluster(**kwargs)
```
I'm a little confused here. What is the hierarchy of execution interfaces? Are some of these parts vestigial? It feels like another layer of abstraction was added, and did not simplify the hierarchy.

It seems like you have the CLI here delegating to `spawn_cluster` which then delegates to `DataflowRunner` but that has multiple modes, one for each kind of execution env? What is the purpose of `run_main` and `cluster_main` then? What does the extra layer of abstraction provided by `DataflowRunner` provide?

I would have assumed that (maybe other than marking `run_main` and `cluster_main` as "private" or "testing" in some way), we could keep `run_main` and `cluster_main` basically untouched, delete `spawn_cluster`, and then all of the logic of parsing command line args to determine execution mode, doing the dynamic `my_dataflow.py` importing, and possibly spawning subprocesses would all happen in just this script. I'm not totally sure I see the utility of the `DataflowRunner` and changing `run_main` and `cluster_main`.

Psykopear replied:
Yes, `DataflowRunner` itself added an extra layer of abstraction that in the end wasn't really adding anything, I think I just got excited with the refactoring :D I removed it, but I left the `WorkerRunner` struct, since that helps us with the bunch of arguments we were passing around to different functions. `run_main` and `cluster_main` are now back to what they were, we lose the ability to raise an exception if they are ran during import though. `spawn_cluster` is the new multiprocess runner. It's implemented on the Rust side just to have a better handling of possible errors and better control of some other aspects like running the webserver, but it can be moved here if we want all of this visible on the python side, the translation is almost 1:1 using `subprocess.Popen`. In fact in the first iteration it was all implemented here and I moved it to the Rust side on a later commit.

Psykopear commented on src/execution/mod.rs:
```diff
@@ -445,204 +447,61 @@ impl PeriodicSpan {
+    dataflow_builder: &str,
+    dataflow_args: Vec<String>,
+) -> PyResult<Dataflow> {
+    let util = py.import("importlib.util")?;
```
Doing the import directly fixes the pickling problem, so I updated the PR following dan's commit on the other branch. I moved the `get_flow` function in the python module (changing the import on the rust side worked too)

Psykopear commented on pysrc/bytewax/inputs.py:
```diff
@@ -53,7 +53,7 @@ def next(self) -> Optional[Any]:
 
         """
-        ...
+        raise NotImplementedError("StatefulSource subclass must implement next")
```
Oh, didn't realize that decorating with `@abstractmethod` together with subclassing `ABC` gave us this behavior, fixed

Psykopear commented on pysrc/bytewax/parse.py:
```diff
@@ -9,60 +9,6 @@
-    return kwargs
-
-
 def proc_env(env: Dict[str, str] = os.environ) -> Dict[str, Any]:
```
yes definitely, pushed a number of commit to add this, the latest ones are fixes done thanks to a review with Esteban

davidselassie commented on src/execution/mod.rs:
```diff
@@ -14,40 +14,50 @@
 //! See [`crate::recovery`] for a description of the recovery
 //! components added to the Timely dataflow.
 
+mod runner;
```
We've also been having the Rust-side module structure parallel the Python-side structure. So if we're going to have `bytewax.run` contain `cluster_main` and (the maybe renamed) `_cli_main`, can this Rust module be renamed to `run` from `execution`?

I do think `run_main` should still be continue to be defined next to `cluster_main`, even though we do publicly expose it in `bytewax.testing` just because they are so similar.

Psykopear replied:
Ok, I renamed the `execution` module to `run`, and only left there the `run` related things (so moved the worker elsewhere, see other comment)

davidselassie commented on src/execution/runner.rs:
```diff
@@ -0,0 +1,240 @@
+
+use super::{build_production_dataflow, PeriodicSpan, WorkerCount, WorkerIndex};
+
+pub(crate) struct WorkerRunner<'a, A: Allocate> {
```
Perhaps call this module `worker`? This is where we define the specifics of the behavior or a worker thread and would better contrast it with the `run` module.

I might also go as far as calling this just `bytewax:worker::Worker` to contrast `timely::worker::Worker`, but I could go either way.

Psykopear replied:
I like `bytewax::worker::Worker`. It shouldn't cause too much confusion since at that point we only use `timely::worker::Worker` in that module, so I moved the `WorkerRunner` and Worker related stuff inside `bytewax::worker`, and renamed `WorkerRunner` to `Worker`.

davidselassie reviewed (commented):
This looks pretty good. I have a few more points of feedback, mostly on naming things so that they either match the Python modules, or reflect their current use.

Psykopear commented on entrypoint.sh:
```diff
@@ -2,7 +2,7 @@
 cd $BYTEWAX_WORKDIR
 . /venv/bin/activate
-python $BYTEWAX_PYTHON_FILE_PATH
+python -m bytewax.run $BYTEWAX_PYTHON_FILE_PATH
```
So, we decided with esteban to keep this name here for compatibility with the helm chart. Also, since the first argument is required in the cli, just setting the env var wouldn't be enough, you still need to pass it as an argument, so the name of the env var itself is not relevant since it's not parsed by our cli. I should remove references to `$BYTEWAX_IMPORT_STR` though

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,237 @@
+    args.import_str = _prepare_import(args.import_str)
+
+    # First of all check if a process_id was set with a different
+    # env var, used in the helm chart for deploy
```
well, this is a custom way we use in the helmchart to define the proc id, which works by stripping the content of a variable from the content of another variable and parsing the resulting string to an int. It's ugly, and I wouldn't want users to do this, it's just a piece of code extracted from the old `parse_env`  for compatibility.

The other one is an env var reprsenting the hostfile path, rather than the addresses themselves. We want users to set the addresses in the variable, but the hostfile path ones was added for the same reason.

I guess at some point we might get rid of the old way of doing things, but Esteban asked me to not break everything for this release, so I added some compatibility workarounds.

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def import_from_string(import_str):
+    if not isinstance(import_str, str):
```
yup, I took this function from Dan's experimental branch, but agree I don't see the use for this, removing it

Psykopear commented on src/execution/mod.rs:
```diff
@@ -849,23 +704,33 @@ pub(crate) fn cluster_main(
-                )
+                    epoch_interval
+                        .clone()
+                        .unwrap_or(EpochInterval::new(Duration::from_secs(10))),
```
Just checked and the `impl Default` is already there, fixing this.

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,237 @@
+__all__ = ["cluster_main"]
+
+
+class ImportFromStringError(Exception):
```
The idea is that this is a user facing error, and it's usually better to see:

```
CustomError: Error while parsing the import string, is it in the right format?
```

rather than a criptic
```
TypeError: NoneType is not a function
```

or something like that. So it is not tagged for catching, but used to wrap possible exceptions into a more comprehensible error for the user. Although the name of the exception is a bit cryptic, I'm going to change it to "BytewaxRunError"

davidselassie replied:
This is solely taste: I do think it's worthwhile to catch and re-raise errors with more descriptive error messages. But I don't think that this needs a custom exception class unless you specifically want to facilitate catching of this class of error programmatically or none of the exception types really describe what's going on. `ValueError` with a good message feels fine to me here. You're welcome to keep it, though.

Psykopear replied:
Makes sense, since we don't need users to catch all the possible errors here, base python exceptions with custom messages will do the job just fine. Fixed

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,237 @@
+        setattr(namespace, self.dest, values)
+
+
+def _prepare_import(import_str):
```
This too was added after Esteban's review to avoid breaking the current helm chart. I don't like it too much, but I took the method from flask's codebase so that at least it behaves in a familiar way for python people, and it seems to work correctly even though I agree it looks fragile and may break in some cases.

Psykopear commented on pysrc/bytewax/run.py:
```diff
@@ -0,0 +1,237 @@
+    try:
+        for attr_str in attrs_str.split("."):
+            if ":" in attr_str:
+                attr_str, _, arg = attr_str.partition(":")
```
Nice, I'm going to look at flask's codebase and try to replicate that then, it makes things clearer than the "double format" I used here for sure

davidselassie commented on CHANGELOG.md:
```diff
@@ -5,6 +5,16 @@
+  prototyping. Production dataflows should be ran by calling the `bytewax.run`
+  module with `python -m bytewax.run <dataflow-path>`.
+  See `python -m bytewax.run -h` for all the possible options.
+  `cluster_main` has been moved to `bytewax.run`, and should be used if you
```
Hmmm. That's a good point. I think we should steer people to use `python -m bytewax.run` and then if they really need Kafka recovery for now, we can point them to `cli_main`. I think we should suggest that Kafka recovery is deprecated.

miccioest reviewed (approved):
Thanks for doing this!
The Bytewax helm chart deploys dataflows running this library version without issues.

Psykopear commented:
Ok, I added 2 more commits to fix the last things:
- Moved `cluster_main` to `bytewax.testing`, since now `cli_main` can do everything `cluster_main` could, so it can be the only officially supported execution funcion
- Removed the custom exception in the `bytewax.run` script
- While fiddling with the `--sqlite-recovery` parameter, I noticed that if you don't pass an existing directory, the flow would panic, so I made the creation of sqlite writers and readers fallible, so that the user sees that there was an error during the construction of the recovery dataflow.
- Updated the CHANGELOG

This was a long one, but it looks we are finally there :smile: Thank you all for the reviews and the patience on this
